lib/auth/challenge_response.rb

```diff
@@ -1,12 +1,22 @@
 require_relative 'authentication_backend'
 
 class ChallengeResponse
-  attr_reader :response_type, :client_id, :session_id, :challenge_name, :challenge_parameters
+  attr_reader :email, :response_type, :client_id, :session_id, :challenge_parameters, :secret_code
+  attr_accessor :challenge_name
 
-  def initialize(cognito_response:)
+  def initialize(email:, cognito_response: nil, secret_code: nil, challenge_name: nil, session: nil, challenge_parameters: nil)
```

Comment:
does email need a default value of `nil`?

Reply:
No we always require email (or username though thats not currently reflected in the code, hence email) so that we can pass round some sort of username.  By the end of the mandatory MFA flow the username ceases to exist so this ensures that it is available at the end of the flow when its needed.